lib/linter/linter.js

```diff
@@ -1537,102 +1556,86 @@ class Linter {
+        const file = new VFile(filenameToExpose, text, {
+            physicalPath: physicalFilename
+        });
+
         const preprocess = options.preprocess || (rawText => [rawText]);
         const postprocess = options.postprocess || (messagesList => messagesList.flat());
+
+        const processorService = new ProcessorService();
+        const preprocessResult = processorService.preprocessSync(file, {
+            processor: {
+                preprocess,
+                postprocess
+            }
+        });
+
+        if (!preprocessResult.ok) {
+            return preprocessResult.errors;
+        }
+
         const filterCodeBlock =
             options.filterCodeBlock ||
             (blockFilename => blockFilename.endsWith(".js"));
         const originalExtname = path.extname(filename);
+        const { files } = preprocessResult;
 
-        let blocks;
-
-        try {
-            blocks = preprocess(text, filenameToExpose);
-        } catch (ex) {
-
-            // If the message includes a leading line number, strip it:
-            const message = `Preprocessing error: ${ex.message.replace(/^line \d+:/iu, "").trim()}`;
-
-            debug("%s\n%s", message, ex.stack);
-
-            return [
-                {
-                    ruleId: null,
-                    fatal: true,
-                    severity: 2,
-                    message,
-                    line: ex.lineNumber,
-                    column: ex.column,
-                    nodeType: null
-                }
-            ];
-        }
-
-        const messageLists = blocks.map((block, i) => {
-            debug("A code block was found: %o", block.filename || "(unnamed)");
+        const messageLists = files.map(block => {
+            debug("A code block was found: %o", block.path || "(unnamed)");
 
             // Keep the legacy behavior.
             if (typeof block === "string") {
                 return this._verifyWithFlatConfigArrayAndWithoutProcessors(block, config, options);
             }
 
-            const blockText = block.text;
-            const blockName = path.join(filename, `${i}_${block.filename}`);
-
             // Skip this block if filtered.
-            if (!filterCodeBlock(blockName, blockText)) {
+            if (!filterCodeBlock(block.path, block.body)) {
                 debug("This code block was skipped.");
                 return [];
             }
 
             // Resolve configuration again if the file content or extension was changed.
-            if (configForRecursive && (text !== blockText || path.extname(blockName) !== originalExtname)) {
+            if (configForRecursive && (text !== block.body || path.extname(block.path) !== originalExtname)) {
                 debug("Resolving configuration again because the file content or extension was changed.");
                 return this._verifyWithFlatConfigArray(
-                    blockText,
+                    block.body,
```

Comment:
Related to https://github.com/eslint/eslint/pull/18818#discussion_r1742470825

Before this change, with this config file:

```js
// eslint.config.js
module.exports = [{
    files: ["**/*.myjs"],
    processor: {
        preprocess(text, filename) {
            return [{ text, filename }];
        },
        postprocess(messages) {
            return [].concat(...messages);
        },
        supportsAutofix: true
    },
    rules: {
        "unicode-bom": ["error", "never"]
    }
}];
```

and this test file (has BOM):

```js
// test.myjs

var x = "This file has BOM";

```

Running `eslint test.myjs` would report that the file has BOM, and running `eslint test.myjs --fix` would remove it.

After this change, `eslint test.myjs` doesn't report that the file has BOM.

Reply:
Oops yeah, that would have to be `rawBody` to make this work.